src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+    public ExtensionMessageHandlerService(IExtensionMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public ValueTask<RegisterExtensionResponse> RegisterExtensionAsync(
+        Solution solution,
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        var analyzerAssemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+
+        CustomMessageHandlerExtension? extension;
+        lock (_lockObject)
+        {
+            // Check if the assembly is already loaded.
+            if (!_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                var analyzerAssemblyLoader = analyzerAssemblyLoaderProvider.CreateNewShadowCopyLoader();
+
+                // Allow this assembly loader to load any dll in assemblyFolderPath.
+                foreach (var dll in Directory.EnumerateFiles(assemblyFolderPath, "*.dll"))
+                {
+                    try
+                    {
+                        // Check if the file is a valid .NET assembly.
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                extension = new CustomMessageHandlerExtension(analyzerAssemblyLoader);
+                _extensions[assemblyFolderPath] = extension;
+            }
+        }
+
+        // AssemblyLoadLockObject is only used to avoid multiple calls from the same extensions to load the same assembly concurrently
+        // resulting in the constructors of the same handlers being called more than once.
+        // All other concurrent operations, including modifying extension.Assemblies are protected by _lockObject.
+        lock (extension.AssemblyLoadLockObject)
+        {
+            if (extension.Assemblies.TryGetValue(assemblyFileName, out var extensionAssembly))
+            {
+                if (extensionAssembly.HasValue)
+                {
+                    return ValueTask.FromResult<RegisterExtensionResponse>(
+                        new(extensionAssembly.Value.WorkspaceMessageHandlers.ToImmutableArray(), extensionAssembly.Value.DocumentMessageHandlers.ToImmutableArray()));
+                }
+                else
+                {
+                    throw new InvalidOperationException($"A previous attempt to load {assemblyFilePath} failed.");
```

Comment:
will this be user visible?  do we need to loc the message?  if so, we can do that later *if* we open an issue tracking this.

Reply:
These will be extender-visible. So I expect them to be localized. I'm happy to localize them now

Reply:
sgtm.

Reply:
ok.  new concern.  it seems like this would not be a place where we want to throw, but rather just log a failure and return that in the response.

put another way InvOpExcept is what we use to indicate a BUG with roslyn code.  An extension not being loadable is not a bug per se, just something we may have to handle given trying to load arbitrary dlls.